runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/JobPreparation.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.fnexecution.jobsubmission;
+
+import com.google.auto.value.AutoValue;
+import com.google.protobuf.Struct;
+import org.apache.beam.model.pipeline.v1.RunnerApi.Pipeline;
+import org.apache.beam.runners.fnexecution.GrpcFnServer;
+import org.apache.beam.runners.fnexecution.artifact.ArtifactStagingService;
+
+/** A job that has been prepared, but not invoked. */
+@AutoValue
+public abstract class JobPreparation {
+  public static Builder builder() {
+    return new AutoValue_JobPreparation.Builder();
+  }
+
+  public abstract String id();
+  public abstract Pipeline pipeline();
+  public abstract Struct options();
+  public abstract GrpcFnServer<ArtifactStagingService> stagingService();
+
+  @AutoValue.Builder
+  abstract static class Builder {
+    abstract Builder setId(String id);
+    abstract Builder setPipeline(Pipeline pipeline);
+    abstract Builder setOptions(Struct options);
+    abstract Builder setStagingService(GrpcFnServer<ArtifactStagingService> stagingService);
```

Comment:
Why is an ArtifactStagingService server part of a JobPreparation? Seems like the server should just live together with the JobService, no?

Reply:
Yeah, ArtifactStagingService is almost certainly the wrong attribute type.  Really this should be the generic artifact reference that we discussed offline.  Once we iron out what that should look like, I'll come back and amend this.

Reply:
key decision: the artifact staging token is the suitable item to return here.  Because of the expected tight coupling between JobService and ArtifactStagingService, a staging token should be enough information for a runner or worker to construct an ArtifactRetrievalService or ArtifactSource on demand.